cli/unipipe/commands/git.ts

```diff
@@ -0,0 +1,147 @@
+import { Command } from "../deps.ts";
+import { Repository } from "../repository.ts";
+
+export interface GitOpts {
+  name?: string;
+  email?: string;
+  message?: string;
+}
+
+export function registerGitCmd(program: Command) {
+  program
+    .command("git <cmd> [repo]")
+    .option(
+      "-n, --name [name:string]",
+      "Git author username. Default is `Unipipe CLI`."
+    )
+    .option(
+      "-e, --email [email:string]",
+      "Git author email. Default is `unipipe-cli@meshcloud.io`."
+    )
+    .option(
+      "-m, --message [message:string]",
+      "Commit message. Default is `Commit changes`."
+    )
+    .description("Runs Git pull/push commands resiliently. It takes care of retrying and rebasing if needed to make sure a push will be successful.")
+    .action(async (opts: GitOpts, cmd: string, repo: string | undefined) => {
+
+      const repository = new Repository(repo ? repo : ".");
+
+      switch (cmd) {
+        case "pull":
+          await commandPull(repository);
+          break;
+        case "push":
+          await commandPush(repository, opts);
+          break;
+        default:
+          console.log(`Git command '${cmd}' is not found`);
+      }
+    });
+}
+
+export async function commandPull(repo: Repository) {
+  const add = await gitAdd(repo.path);
+  if (!add) return;
```

Comment:
d: Just to make sure i get it right, the `if (!add) return;` handles the case that something went wrong during gitAdd? So it failed with an error? If that is the case, it is currently a bit hard to understand as just reading the code here it seems like "if there is nothing to add, then return". But that would be wrong, right? Because we want to pull changes if there are no local changes, right? Perhaps you can make the code a bit more expressive around this?

Reply:
Yes. This is just a status variable used to terminate execution if an error occurs. At the same time, full logs and an error message will be printed into the console. For example 'git add .' can fail when git repository is not initiated. 
In this case, the output will be like this:

> Running git add .
fatal: not a git repository (or any of the parent directories): .git
Command git add . failed

I'll rename variables to add 'Successful' suffix to make them more clear.